add NFIR package

tomas-krupa opened the pull request:
### Summary
NIST-made library and tool **NFIR/0.2.0** conanized

#### Motivation
NFIR is a library (and tool) capable of upsampling or downsampling fingerprint images according to NIST standards. It can now be seamlessly integrated as a Conan dependency.

#### Details
Since the official NIST repository does not contain any Releases, I had to fork the repository and create a Release myself in order to link the Conan sources to it.

---
- [x] Read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md)
- [x] Checked that this PR is not a duplicate: [list of PRs by recipe](https://github.com/conan-io/conan-center-index/discussions/24240)
- [x] Tested locally with at least one configuration using a recent version of Conan

tomas-krupa committed:
add NFIR package

valgur commented on recipes/nfir/all/conandata.yml:
```diff
@@ -0,0 +1,11 @@
+sources:
+  "0.2.0":
+    url: "https://codeload.github.com/tomas-krupa/NFIR/zip/refs/tags/0.2.0"
+    sha256: "836ad87ae6a25c9a7f361a854d30c25560e7af833a5daeb906e4867f93b16276"
```
```suggestion
    url: "https://github.com/usnistgov/NFIR/archive/b670b673f8c397f37b231d16bc8b41f934d05841.tar.gz"
    sha256: "0bce5cd844560f82968540b0ae27248973bc1f3bfccadfc80e687f129a5588e8"
```
You can use the latest commit from the original repo and set the version to `0.2.0.cci.20241122` to account for no exact releases being made.

valgur commented on recipes/nfir/all/conanfile.py:
```diff
@@ -0,0 +1,56 @@
+    license = "NIST"
+    description = "The NIST Fingerprint Image Resampler NFIR is a library capable of upsampling or downsampling fingerprint images."
+    settings = "os", "compiler", "build_type", "arch"
+    options = { "shared": [True, False] }
+    default_options = { "shared": False }
```
```suggestion
    options = { "shared": [True, False], "fPIC": [True, False] }
    default_options = { "shared": False, "fPIC": True }
```
`fPIC` option needs to be provided for `library` package type.

valgur commented on recipes/nfir/all/conanfile.py:
```diff
@@ -0,0 +1,56 @@
+    default_options = { "shared": False }
+    generators = "CMakeDeps"
+    package_type = "library"
+    exports_sources = "CMakeLists.txt", "patches/*"
```
```suggestion
    implements = ["auto_shared_fpic"]
```
`exports_sources` is unnecessary on CCI. `auto_shared_fpic` is necessary for the correct handling of `shared` and `fPIC` options: https://docs.conan.io/2/reference/conanfile/methods/configure.html#auto-shared-fpic

valgur commented on recipes/nfir/all/conanfile.py:
```diff
@@ -0,0 +1,56 @@
+        tc.variables["USE_NFIMM"] = False
+        tc.variables["_WIN32_64"] = self.settings.os == "Windows"
+        tc.generate()
+
```
```suggestion
        deps = PkgConfigDeps(self)
        deps.generate()

```
pkg-config is used to find OpenCV https://github.com/usnistgov/NFIR/blob/b670b673f8c397f37b231d16bc8b41f934d05841/src/lib/CMakeLists.txt#L46

tomas-krupa replied:
just adding this does not work on windows, and it also introduces unnecessary tool requirements

valgur commented on recipes/nfir/all/conanfile.py:
```diff
@@ -0,0 +1,56 @@
+    def package_info(self):
+        self.cpp_info.libs = ["NFIR_ITL"]
+        self.cpp_info.set_property("cmake_file_name", "NFIR")
+        self.cpp_info.set_property("cmake_target_name", f"NFIR::NFIR")
```
```suggestion
        self.cpp_info.set_property("cmake_target_name", "NFIR::NFIR")
```

valgur commented on recipes/nfir/all/patches/0001-find-OpenCV-provided-by-conan_toolchain.cmake.patch:
```diff
@@ -0,0 +1,65 @@
++find_package(OpenCV REQUIRED CONFIG)
++target_link_libraries(${PROJECT_NAME} PRIVATE opencv::opencv)
+ 
+ if(_WIN32_64)
```
Simply replace `_WIN32_64` with `FALSE` here. The non-Windows logic is valid for other platforms as well and `/WX` is generally unwanted in public builds anyway.

valgur commented on recipes/nfir/all/patches/0002-enable-choose-between-shared-and-static-library.patch:
You can most likely drop this patch. Conan automatically sets `BUILD_SHARED_LIBS` which CMake uses to determine whether STATIC or SHARED should be built.

valgur commented on recipes/nfir/all/patches/0004-set-NFIR-executable-name-to-NFIR_bin-tool.patch:
Can probably be dropped or at least replaced with `install()` calls to make `cmake.install()` work.

valgur commented on recipes/nfir/all/patches/0003-add-NFIR-resample-overload.patch:
Patches for functional changes to the library are not going to be accepted on CCI, unfortunately. You should submit it as a PR to the upstream project instead.

tomas-krupa committed:
fixed code review findings

tomas-krupa commented on recipes/nfir/all/conanfile.py:
```diff
@@ -0,0 +1,56 @@
+        bindir = path.join(self.package_folder, "bin")
+        self.runenv_info.append_path("PATH", bindir)
+        self.buildenv_info.append_path("PATH", bindir)
+        self.env_info.PATH.append(bindir)
```
without  ``self.runenv_info.append_path("PATH", bindir) i get``

```
nfir/0.2.0.cci.20241122 (test package): RUN: NFIR -a 1000 -b 500 -c image.png -d downsampled.png
/bin/sh: 1: NFIR: not found
```

tomas-krupa commented:
@valgur, thank you for the helpful feedback. I'm curious, the PR was reviewed very quickly after being opened. However, I have two other PRs that have been waiting for review for several weeks/months with no activity. Is there any official way I can request a review for a PR?

tomas-krupa committed:
fixed linter issues

tomas-krupa committed:
fix conan CI build issues

valgur commented:
> However, I have two other PRs that have been waiting for review for several weeks/months with no activity. Is there any official way I can request a review for a PR?

That's just how things work down here, I guess. :neutral_face:

I'm just a regular contributor, and don't have direct bearing on whether this gets merged or not, but I hope my feedback helps things move along a bit faster at least.

tomas-krupa committed:
nfir: fix NFIR executable not found in test_package on macOS

Add VirtualRunEnv to generators so Conan produces the conanrun.sh script
that exports the package bin/ directory into PATH, allowing
to resolve correctly on macOS.

tomas-krupa committed:
nfir: exercise resample() API in test_package

Replace the trivial printVersion() call with an actual resample()
invocation using a minimal in-memory 1x1 PGM image (P5 binary format),
upsampling from 300 to 500 PPI. This verifies that the core library
function and its OpenCV dependency are correctly linked and functional.
